### PR TITLE
Decouple ClassType definitions and Lite interpreter.

### PR DESCRIPTION
**Pull request by iseeyuan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27786 Decouple ClassType definitions and Lite interpreter.**

Differential Revision: [D17909998](https://our.internmc.facebook.com/intern/diff/D17909998)

The lite interpreter depends on ClassType definitions, which depends on jit::Function and then most of JIT codes. This PR is to remove this dependency. After this decoupling we could remove most of the JIT dependecies from lite interpreter.

As a sibling class to ClassType, MobileType is introduced and returned to class_resolver. Its implementation can be kept separately from ClassType's implementation. 

As an example, for an internal build, the interpreter binary size (without any build optimization) can be reduced from 28.4 MB to 7.5 MB.

### CODE REVIEW

File: torch/csrc/jit/mobile/import.cpp [hunk 1]
```diff
@@ -149,12 +149,8 @@ c10::IValue BytecodeDeserializer::readArchive(const std::string& archive_name) {
   };
 
   auto class_resolver = [&](const c10::QualifiedName& qn) {
-    if (compilation_unit_->get_class(qn) == nullptr) {
```
Comment (zdevito) [inline torch/csrc/jit/mobile/import.cpp]: Looking at the APIs closely, I think we should make the following changes:
* class_resolver_ returns TypePtr, not StrongTypePtr
* obj_loader_ takes TypePtr, not StrongTypePtr
* the implementation of obj_loader_ can construct a StrongTypePtr because it can have a reference to the compilation unit. 

This enables us to keep all the standard asserts in StrongTypePtr while still allowing mobile to leave the type null.

File: torch/csrc/jit/mobile/import.cpp [hunk 2]
```diff
@@ -150,15 +150,15 @@ c10::IValue BytecodeDeserializer::readArchive(const std::string& archive_name) {
...
-      auto typeptr = ClassType::create(qn, compilation_unit_, true);
-      compilation_unit_->register_type(typeptr);
-    }
-    return c10::StrongTypePtr(
-        compilation_unit_, compilation_unit_->get_class(qn));
+    return nullptr;
   };
 
-  auto obj_loader = [&](at::StrongTypePtr type, IValue input) {
+  auto obj_loader = [&](c10::ClassTypePtr typePtr, IValue input) {
```
Comment (zdevito) [inline torch/csrc/jit/mobile/import.cpp]: Probaby don't want to create a new class type for every object. It would be better to a have a single class type for everything so there are not a lot of different "dummy" types floating around.

File: torch/csrc/jit/unpickler.cpp [hunk 3]
```diff
@@ -59,6 +59,8 @@ void restoreAccurateTypeTags(const IValue& root, const TypePtr& type_tag) {
   while (!to_process.empty()) {
     Work w = std::move(to_process.back());
     to_process.pop_back();
+    if (!w.static_type)
```
Comment (zdevito) [inline torch/csrc/jit/unpickler.cpp]: The type is not null in the current implementation, so why is this needed?

**Review by zdevito — COMMENTED** [review]
This seems to make mobile rely on having access to ClassType still to create the dummy class. I thought there was a requirement to avoid that.

File: torch/csrc/jit/import_legacy.cpp [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 #include "caffe2/core/common.h"
 #include "caffe2/core/types.h"
-#include "caffe2/proto/caffe2_pb.h"
+#include <caffe2/proto/caffe2_pb.h>
```
Comment (pietern) [inline torch/csrc/jit/import_legacy.cpp]: Why? This shouldn't be needed -- it's not a system include.

Comment (iseeyuan) [reply torch/csrc/jit/import_legacy.cpp]: I was trying to remove this include to resolve a tidy issue and added it back because some caffe2 test failure. I'll change the style back. Thanks!

**Review by zdevito — COMMENTED** [review]
The way this forces Object to no longer be able to return a ClassType object is very intrusive to existing code. It makes sense that a object will return a class type.  I think we are going to need to figure out a way for classtype to exist in the mobile build by cutting off the functionality of ClassType at another point.
